src/tasks/Tasks.ts

```diff
@@ -0,0 +1,869 @@
+      schedulingLoop,
+      abortController
+    );
+    this.schedulerLogger.info('Started Scheduling Loop');
+  }
+
+  protected async startQueueing() {
+    if (this.queuingLoop != null) return;
+    this.queueLogger.info('Starting Queueing Loop');
+    const abortController = new AbortController();
+    const abortP = utils.signalPromise(abortController.signal);
+    const queuingLoop = (async () => {
+      while (!abortController.signal.aborted) {
+        try {
+          await Promise.race([this.queuingLock.waitForUnlock(), abortP]);
+        } catch (e) {
+          if (e === abortQueuingLoopReason) {
+            break;
+          } else {
+            throw e;
+          }
+        }
+        this.queueLogger.debug(`Begin queuing loop iteration`);
+        [this.queuingLockReleaser] = await this.queuingLock.lock()();
+        await this.db.withTransactionF(async (tran) => {
+          for await (const [kP, taskIdBuffer] of tran.iterator(
+            this.tasksQueuedDbPath,
+          )) {
+            if(abortController.signal.aborted) break;
+            // Remember the activePromises isn't being incremented
+            if (this.activePromises.size >= this.activeLimit) break;
+            // Remove task from the queued index
+            await tran.del([...this.tasksQueuedDbPath, ...kP]);
+            const taskId = IdInternal.fromBuffer<TaskId>(taskIdBuffer);
+            const taskData = (await tran.get<TaskData>([...this.tasksTaskDbPath, taskIdBuffer]))!;
+            this.queueLogger.debug(`Queueing loop iteration starting Task ${taskId.toMultibase('base32hex')}`);
+            await this.startTask(taskId, taskData);
+          }
+        });
+        this.queueLogger.debug(`Finish queuing loop iteration`);
+      }
+    })();
+    // Cancellation is always a resolution
+    // the promise must resolve, by waiting for resolution
+    // it's graceful termination of the loop
+    this.queuingLoop = PromiseCancellable.from(
+      queuingLoop,
+      abortController
+    );
+    this.queueLogger.info('Started Queueing Loop');
+  }
+
+  protected async stopScheduling (): Promise<void> {
+    if (this.schedulingLoop == null) return;
+    this.logger.info('Stopping Scheduling Loop');
+    // Cancel the timer if it exists
+    this.schedulingTimer?.cancel();
+    // Cancel the scheduling loop
+    this.schedulingLoop.cancel(abortSchedulingLoopReason);
+    // Wait for the cancellation signal to resolve the promise
+    await this.schedulingLoop;
+    // Indicates that the loop is no longer running
+    this.schedulingLoop = null;
+    this.logger.info('Stopped Scheduling Loop');
+  }
+
+  protected async stopQueueing() {
+    if (this.queuingLoop == null) return;
+    this.logger.info('Stopping Queuing Loop');
+    // Cancel the scheduling loop
+    this.queuingLoop.cancel(abortQueuingLoopReason);
+    await this.queuingLoop;
+    this.queuingLoop = null;
+    this.logger.info('Stopped Queuing Loop');
+  }
```

Comment:
We need to wait for any active promises inside `this.activePromises` to resolve/reject after cancelling. Otherwise we will end up with hard to debug `ErrorXNotStarted` race condition errors.

Reply:
Right now `stopQueuing` only has 1 responsibility, stopping the queueing. Nothing else.

A different function should be used to `stopTasks`. However once we wrap the tasks as `PromiseCancellable`, this should be doable, by simply signalling to all the tasks to cancel in the active promise list.

Some thing like:

```
for (const p of this.activePromises) {
  p.cancel();
}
await Promise.any([...this.activePromises.getValues()]);
```

Pseudo code above ^.

Reply:
I believe we have a function called `startTask`.

Symmetrically one can have a function called `stopTask`.

Then have another function called `stopTasks` that the `Tasks.stopProcessing` should call.

They can all be done in parallel so that's why I used `Promises.all([])`.

Reply:
Note that we should be using lazy cancelling not eager cancelling. Because we expect all tasks to gracefully stop.

Reply:
The `stopTask` should just do promise cancellation. It should not be affecting the database in any way. Let the `finally` handler do any database operations, because it ends up calling `gcTask`.

Reply:
I think this has been resolved.